### DISCUSSION TITLE
Federated website edits by pull request

### DISCUSSION THREAD

**Discussion by erlend-sh** [muni-town/weird#15] [Ideas]
This is by no means a core feature of Weird, but it might be something we can help facilitate for people's websites by letting them easily add a widget that enables this type of interaction.

Imagine I've got a regular website, with a regular blog, nothing special. I do however also have an ActivityPub account.

I would like to enable pull requests for my blog posts, primarily for the purpose of typo edits, although more in-depth collaborative applications are very conceivable here as well.

To do that, I'm gonna go to `edit.me` (not a real website or final domain yet) and authenticate with my fediverse account. Then I'm gonna give the site the rss feed of my blog, so it knows that those articles should be made available for community editing.

<details>
  <summary>Same auth flow as fedview for example</summary>
https://fediview.com/
<img width="823" alt="Screenshot 2024-04-11 at 22 46 08" src="https://github.com/commune-os/weird/assets/583842/2df704dc-b188-4a8c-b3ba-c61f739716d0">
</details>

The site is gonna give me a little code snippet (maybe a Web Component?) for me to insert into my blog. There it'll show up as a as a pencil ✏️ icon widget in the bottom corner. When someone else clicks on this widget, it'll ask them if they want to suggest an edit to this page. If they click onwards, they'll be taken to that blog post's edit page, e.g. `edit.me/blog.erlend.sh/aj73halh` where the page has been imported as text-only.

Before they can edit the page they'll be asked to authenticate with their fedi account. Upon submitting their edit, two things will happen:

- A new revision of the page has been saved (maybe we do this via CRDT?), with the changes credited to the new editor.
- The fedi account of `edit.me` will send a DM to the site owner and the edit-contributor to notify both parties that a pull request has been sent and is pending review.

The site owner can go to `edit.me` to accept the PR. Now, that doesn't actually change the source page. However, once the site owner has gone to the source page and edited it accordingly, the newest sync on `edit.me` will reflect whether or not those changes have indeed been incorporated. And by having accepted the PR, the edit-contributor's [proof of work](https://blog.erlend.sh/socialized-proof-of-work) is recorded.

---

It's a little clunky, but it's the best I can think of for a setup that accommodates *any* kind of website, rather than requiring everyone to be on the same system in order to make PRs happen.

**Comment by aschrijver** [comment]
I might add to this idea that the [ForgeFed](https://forgefed.org) specification details with how PR's (Merge Requests) are handled as an ActivityPub extension. While the focus of the project is forge federation, there's nothing that says that both ends of the federation must be forges.

In a [recent blog post](https://forgefed.org/blog/teams/) the status of the project is outlined. The Vervis reference implementation now has support for Projects and Teams, that may play a role in site maintenance as well, and there's an implementation of Object Capabilities to deal with permissions/authorizations (Grants).
